src/Interpreters/Context.cpp

```diff
@@ -747,15 +747,17 @@ VolumePtr Context::getTemporaryVolume() const
 
 TemporaryDataOnDiskScopePtr Context::getTempDataOnDisk() const
 {
-    auto lock = getLock();
     if (this->temp_data_on_disk)
         return this->temp_data_on_disk;
+
+    auto lock = getLock();
     return shared->temp_data_on_disk;
 }
 
 void Context::setTempDataOnDisk(TemporaryDataOnDiskScopePtr temp_data_on_disk_)
 {
-    auto lock = getLock();
+    /// It's set from `ProcessList::insert` in `executeQueryImpl` before query execution
+    /// so no races with `getTempDataOnDisk` which is called from query execution.
     this->temp_data_on_disk = std::move(temp_data_on_disk_);
```

Comment:
What about other functions that use `temp_data_on_disk`? For example
https://github.com/ClickHouse/ClickHouse/blob/714b54b322c9d7908d32c64f17f7d3e1cd0981b9/src/Interpreters/Context.cpp#L739-L743

Reply:
Oh, we have two different `temp_data_on_disk` one in shared part of context and one in current 🤔  I'll check why do we need two and when it was introduced. Probably shared is global and in current context is for query execution. Global one is set on server startup https://github.com/ClickHouse/ClickHouse/blob/ceef48a1627e75ca780692894724f7f170dac4bf/programs/server/Server.cpp#L1048-L1060

Anyway, better naming and comments are required, so I'm going to investigate how to make it better and continue this PR

Reply:
@tavplubix 
I renamed `temp_data_on_disk` in shared context and added comments.
Also added locks for get/set this variable in *shared* context (not sure if it's mandatory since it's set once at `Server::main`), but in other places accessing shared parts of context we do (e.g. `Context::setFlagsPath`)